### PR TITLE
timer: fix oss-fuzz issue #11852

### PR DESCRIPTION
**Pull request by ipuustin**
Description:

Fix a time conversion signed int overflow (https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=11852).

Risk Level: low
Testing: fuzz

### CODE REVIEW

File: test/server/server_corpus/clusterfuzz-testcase-minimized-server_fuzz_test-5686444035670016 [hunk 1]
```diff
@@ -0,0 +1,2 @@
+static_resources {   clusters {     name: " "     connect_timeout {       nanos: 4     }     hosts {
```
Comment (htuch) [inline test/server/server_corpus/clusterfuzz-testcase-minimized-server_fuzz_test-5686444035670016]: Could you also add a unit test for this new clamping or conversion boundary behavior somewhere, e.g. https://github.com/envoyproxy/envoy/blob/master/test/common/event/dispatcher_impl_test.cc#L189? The corpus entries are useful, but they are hard to reason about when trying to understand behavior, and are subject to fuzzer rot.

**Comment by htuch** [comment]
@ipuustin friendly ping on this one.

File: source/common/event/timer_impl.cc [hunk 2]
```diff
@@ -23,7 +23,17 @@ void TimerImpl::enableTimer(const std::chrono::milliseconds& d) {
     event_active(&raw_event_, EV_TIMEOUT, 0);
   } else {
     // TODO(#4332): use duration_cast more nicely to clean up this code.
-    std::chrono::microseconds us = std::chrono::duration_cast<std::chrono::microseconds>(d);
+
```
[the remark this answers is not in the document]
Comment (ipuustin) [reply source/common/event/timer_impl.cc]: The method described in the TODO seems to work fine, and removes the need for clamping the value. Using that now!

File: test/server/server_corpus/clusterfuzz-testcase-minimized-server_fuzz_test-5686444035670016 (hunk 1 above)
Comment (ipuustin) [reply test/server/server_corpus/clusterfuzz-testcase-minimized-server_fuzz_test-5686444035670016]: I added some tests. I refactored the timeval conversion to a separate function to make testing easier.

File: source/common/event/timer_impl.cc [hunk 3]
```diff
@@ -18,25 +18,21 @@ TimerImpl::TimerImpl(Libevent::BasePtr& libevent, TimerCb cb) : cb_(cb) {
 
 void TimerImpl::disableTimer() { event_del(&raw_event_); }
 
+void TimerImpl::millisecondsToTimeval(timeval* tv, const std::chrono::milliseconds& d) {
```
Comment (htuch) [inline source/common/event/timer_impl.cc]: Tiny nit: Envoy tends to have the order be input parameters followed by output parameters, and to use mutable ref. So, can this be `TimerImpl::millisecondsToTimeval(const std::chrono::milliseconds& d, timeval& tv)`?

File: source/common/event/timer_impl.cc [hunk 4]
```diff
@@ -18,25 +18,21 @@ TimerImpl::TimerImpl(Libevent::BasePtr& libevent, TimerCb cb) : cb_(cb) {
...
+  ASSERT(tv);
```
Comment (htuch) [inline source/common/event/timer_impl.cc]: This can be skipped

File: source/common/event/timer_impl.h [hunk 5]
```diff
@@ -22,6 +22,9 @@ class TimerImpl : public Timer, ImplBase {
   void enableTimer(const std::chrono::milliseconds& d) override;
   bool enabled() override;
 
+  // Public for testing.
+  void millisecondsToTimeval(timeval* tv, const std::chrono::milliseconds& d);
```
Comment (htuch) [inline source/common/event/timer_impl.h]: Suggest putting in a utility class, e.g. `TimerUtils` and making `static`

File: source/common/event/timer_impl.cc (hunk 3 above)
Comment (ipuustin) [reply source/common/event/timer_impl.cc]: Sure! This is good advice regarding the code conventions.

File: source/common/event/timer_impl.h (hunk 5 above)
Comment (ipuustin) [reply source/common/event/timer_impl.h]: Ok. I was originally aiming to do "minimal" changes but a separate class makes sense.
